fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/FDBRecordStore.java

```diff
@@ -2152,13 +2154,18 @@ public Range getUnbuiltRange() {
                                                                                   @Nonnull IsolationLevel storeHeaderIsolationLevel,
                                                                                   @Nonnull IsolationLevel indexStateIsolationLevel) {
         CompletableFuture<RecordMetaDataProto.DataStoreInfo> storeHeaderFuture = loadStoreHeaderAsync(existenceCheck, storeHeaderIsolationLevel);
-        CompletableFuture<Map<String, IndexState>> loadIndexStates = loadIndexStatesAsync(indexStateIsolationLevel);
-        return context.instrument(FDBStoreTimer.Events.LOAD_RECORD_STORE_STATE, storeHeaderFuture.thenCombine(loadIndexStates, RecordStoreState::new));
+        CompletableFuture<Map<String, IndexState>> loadIndexStates = storeHeaderFuture.thenCompose(vignore -> loadIndexStatesAsync(indexStateIsolationLevel, storeHeaderFuture.join()));
+        return context.instrument(FDBStoreTimer.Events.LOAD_RECORD_STORE_STATE, loadIndexStates.thenApply(vignore -> new RecordStoreState(storeHeaderFuture.join(), loadIndexStates.join())));
     }
```

Comment:
Yeah, this is a valid point. Arguably performance on this path really matters (this is called almost on every record store open). While writing this, I thought about it but I don't know how better this can be. I like the idea of reading from the new key space speculatively, but I'm not sure how to tell if it's right or wrong.

Reply:
To clarify, I understood what you meant by speculatively loading the states. I guess my concern was how to choose the right starting location. If the starting location is chosen poorly, we will end up loading the state twice a lot of times. Sorry for ambiguity. 

> The other approach 

I think I still prefer keeping them separate because of safety issues that we discussed before.